File: generic/Color.cpp
#include "Color.h"

Color::Color() {
    r = g = b = 255;
}

Color::Color(unsigned char red, unsigned char green, unsigned char blue) {
    Set(red, green, blue);
}

void Color::Set(unsigned char red, unsigned char green, unsigned char blue) {
    r = red;
    g = green;
    b = blue;
}

double Color::Distance(Color const &e1, Color const &e2) {
    long rmean = ((long)e1.r + (long)e2.r) / 2;
    long r = (long)e1.r - (long)e2.r;
    long g = (long)e1.g - (long)e2.g;
    long b = (long)e1.b - (long)e2.b;
    return sqrt((((512 + rmean)*r*r) >> 8) + 4 * g*g + (((767 - rmean)*b*b) >> 8));
}

Color &Color::operator=(Color const &rhs) {
    r = rhs.r;
    g = rhs.g;
    b = rhs.b;
    return *this;
}

bool Color::operator== (Color const &rhs) const {
    return r == rhs.r && g == rhs.g && b == rhs.b;
}

bool Color::operator!= (Color const &rhs) const {
    return r != rhs.r || g != rhs.g || b != rhs.b;
}

unsigned int Color::FindIndexInTable(std::vector<std::pair<unsigned int, Color>> const &table, unsigned int tableMin, unsigned int tableMax) const {
    if (tableMax == 0)
        tableMax = 9999;
    for (auto &clr : table) {
        if (clr.first >= tableMin && clr.first <= tableMax) {
            if (*this == clr.second)
                return clr.first;
        }
    }
    unsigned int result = 0;
    double dist = 9999.0;
    for (auto &clr : table) {
        if (clr.first >= tableMin && clr.first <= tableMax) {
            auto newDist = Distance(*this, clr.second);
            if (newDist < dist) {
                dist = newDist;
                result = clr.first;
            }
        }
    }
    return result;
}

void Color::SetFromTable(std::vector<std::pair<unsigned int, Color>> const &table, unsigned int index, unsigned int tableMin, unsigned int tableMax) {
    if (!table.empty()) {
        if (tableMax == 0)
            tableMax = 9999;
        if (index >= tableMin && index <= tableMax) {
            for (auto &clr : table) {
                if (clr.first == index) {
                    *this = clr.second;
                    return;
                }
            }
        }
    }
    Set(255, 255, 255);
}

std::wstring Color::ToStr() const {
    return Utils::Format(L"(%u,%u,%u)", r, g, b);
}

std::wstring Color::ToHexStr() const {
    return Utils::Format(L"#%02X%02X%02X", r, g, b);
}

ColorPair::ColorPair() {}

ColorPair::ColorPair(Color const &color1, Color const &color2) {
    Set(color1, color2);
}

void ColorPair::Set(Color const &color1, Color const &color2) {
    first = color1;
    second = color2;
}

double ColorPair::Distance(ColorPair const &e1, ColorPair const &e2) {
    auto dist1 = (Color::Distance(e1.first, e2.first) + Color::Distance(e1.second, e2.second)) / 2.0;
    auto dist2 = (Color::Distance(e1.first, e2.second) + Color::Distance(e1.second, e2.first)) / 2.0;
    return fmin(dist1, dist2);
}

ColorPair &ColorPair::operator=(ColorPair const &rhs) {
    first = rhs.first;
    second = rhs.second;
    return *this;
}

bool ColorPair::operator== (ColorPair const &rhs) const {
    return (first == rhs.first && second == rhs.second) || (first == rhs.second && second == rhs.first);
}

unsigned int ColorPair::FindIndexInTable(std::vector<std::pair<unsigned int, ColorPair>> const &table, unsigned int tableMin, unsigned int tableMax) const {
    if (tableMax == 0)
        tableMax = 9999;
    for (auto &clr : table) {
        if (clr.first >= tableMin && clr.first <= tableMax) {
            if (*this == clr.second)
                return clr.first;
        }
    }
    unsigned int result = 0;
    double dist = 9999.0;
    for (auto &clr : table) {
        if (clr.first >= tableMin && clr.first <= tableMax) {
            auto newDist = Distance(*this, clr.second);
            if (newDist < dist) {
                dist = newDist;
                result = clr.first;
            }
        }
    }
    return result;
}

void ColorPair::SetFromTable(std::vector<std::pair<unsigned int, ColorPair>> const &table, unsigned int index, unsigned int tableMin, unsigned int tableMax) {
    if (!table.empty()) {
        if (tableMax == 0)
            tableMax = 9999;
        if (index >= tableMin && index <= tableMax) {
            for (auto &clr : table) {
                if (clr.first == index) {
                    *this = clr.second;
                    return;
                }
            }
        }
    }
    Set(Color(), Color());
}

std::wstring ColorPair::ToStr() const {
    return Utils::Format(L"%s:%s", first.ToStr().c_str(), second.ToStr().c_str());
}

std::wstring ColorPair::ToHexStr() const {
    return Utils::Format(L"%s:%s", first.ToHexStr().c_str(), second.ToHexStr().c_str());
}

File: tools/kit_comparison_tool/color.py
import math
from typing import List, Tuple, Union
from colormath.color_objects import sRGBColor, LabColor
from colormath.color_conversions import convert_color
from colormath.color_diff import delta_e_cie2000
import numpy

def patch_asscalar(a):
    return a.item()
    
setattr(numpy, "asscalar", patch_asscalar)

class Utils:
    @staticmethod
    def format(fmt: str, *args):
        return fmt % args

class Color:
    def __init__(self, red: Union[int, None] = None, green: int = None, blue: int = None):
        if red is None:
            self.r, self.g, self.b = 255, 255, 255
        elif green is None and blue is None:
            self.set_from_int(red)
        else:
            self.set(red, green, blue)

    def set(self, red: int, green: int, blue: int):
        self.r, self.g, self.b = red, green, blue

    def set_from_int(self, int_value: int):
        self.r = (int_value >> 16) & 0xFF
        self.g = (int_value >> 8) & 0xFF
        self.b = int_value & 0xFF

    @staticmethod
    def distance(e1: 'Color', e2: 'Color') -> float:
        srgb1 = sRGBColor(e1.r / 255.0, e1.g / 255.0, e1.b / 255.0)
        srgb2 = sRGBColor(e2.r / 255.0, e2.g / 255.0, e2.b / 255.0)
        lab1 = convert_color(srgb1, LabColor)
        lab2 = convert_color(srgb2, LabColor)
        return delta_e_cie2000(lab1, lab2)

    def __eq__(self, other: 'Color') -> bool:
        return self.r == other.r and self.g == other.g and self.b == other.b

    def __ne__(self, other: 'Color') -> bool:
        return not self == other

    def __lt__(self, other: 'Color') -> bool:
        return self.to_hex_str() <= other.to_hex_str()

    def find_index_in_table(self, table: List[Tuple[int, 'Color']], table_min=0, table_max=0) -> int:
        if table_max == 0:
            table_max = 9999
        for index, color in table:
            if table_min <= index <= table_max and self == color:
                return index

        closest_index = 0
        min_dist = float('inf')
        for index, color in table:
            if table_min <= index <= table_max:
                dist = self.distance(self, color)
                if dist < min_dist:
                    min_dist = dist
                    closest_index = index
        return closest_index

    def set_from_table(self, table: List[Tuple[int, 'Color']], index: int, table_min=0, table_max=0):
        if not table:
            self.set(255, 255, 255)
            return

        if table_max == 0:
            table_max = 9999

        for idx, color in table:
            if idx == index and table_min <= idx <= table_max:
                self.r, self.g, self.b = color.r, color.g, color.b
                return

        self.set(255, 255, 255)

    def to_str(self) -> str:
        return Utils.format("(%u,%u,%u)", self.r, self.g, self.b)

    def to_hex_str(self) -> str:
        return Utils.format("#%02X%02X%02X", self.r, self.g, self.b)

    def to_int(self) -> int:
        return (0xFF << 24) | (self.r << 16) | (self.g << 8) | self.b


GenColor = Color


class ColorPair:
    def __init__(self, color1: Color = None, color2: Color = None):
        if color1 is None:
            self.first = Color()
            self.second = Color()
        else:
            self.set(color1, color2)

    def set(self, color1: Color, color2: Color):
        self.first = color1
        self.second = color2

    @staticmethod
    def distance(e1: 'ColorPair', e2: 'ColorPair') -> float:
        d11 = Color.distance(e1.first, e2.first)
        d22 = Color.distance(e1.second, e2.second)
        d12 = Color.distance(e1.first, e2.second)
        d21 = Color.distance(e1.second, e2.first)
        best_match = min(d11, d22, d12, d21)
        average_all = (d11 + d22 + d12 + d21) / 4.0
        return (best_match + average_all) / 2.0

    def __eq__(self, other: 'ColorPair') -> bool:
        return ((self.first == other.first and self.second == other.second) or
                (self.first == other.second and self.second == other.first))

    def find_index_in_table(self, table: List[Tuple[int, 'ColorPair']], table_min=0, table_max=0) -> int:
        if table_max == 0:
            table_max = 9999
        for index, pair in table:
            if table_min <= index <= table_max and self == pair:
                return index

        closest_index = 0
        min_dist = float('inf')
        for index, pair in table:
            if table_min <= index <= table_max:
                dist = ColorPair.distance(self, pair)
                if dist < min_dist:
                    min_dist = dist
                    closest_index = index
        return closest_index

    def set_from_table(self, table: List[Tuple[int, 'ColorPair']], index: int, table_min=0, table_max=0):
        if not table:
            self.set(Color(), Color())
            return

        if table_max == 0:
            table_max = 9999

        for idx, pair in table:
            if idx == index and table_min <= idx <= table_max:
                self.first = pair.first
                self.second = pair.second
                return

        self.set(Color(), Color())

    def to_str(self) -> str:
        return f"{self.first.to_str()}:{self.second.to_str()}"

    def to_hex_str(self) -> str:
        return f"{self.first.to_hex_str()}:{self.second.to_hex_str()}"


GenColorPair = ColorPair
